src/solver.hs

```diff
@@ -0,0 +1,232 @@
+module Solver where
+
+import Sudoku
+import SudokuLogic
+import Sudokus
+import Optimizer
+import Data.List
+data Step =     LastFreeCellBlock Position      |
+                LastFreeCellRow Position        |
+                LastFreeCellCollumn Position    |
+                SingleCandidate Position        |
+                SinglePositionRow Position      |
+                SinglePositionColumn Position   |
+                SinglePositionBlock Position    |
+                CandidateLine Position          |
+                NakedPairRow Position           |
+                NakedPairColumn Position        |
+                NakedPairBlock Position         |
+                HiddenPairRow Position          |
+                HiddenPairColumn Position       |
+                HiddenPairBlock Position        |
+                NOAVAILABLESTEPS
+    deriving (Eq,Show) -- | Then add more step types
+
+type Solution = [Step]
+
+
+-- | Takes a sudoku and returns a reorginised list of steps
+optimSteps :: Sudoku -> [Position -> Step]
+optimSteps sud = sudStepOrder $ zip (stepWeight sud) steps
+
+-- | List of every step the solver can try
+steps :: [Position -> Step]
+steps = [--LastFreeCellBlock, LastFreeCellRow, LastFreeCellCollumn, 
+        SingleCandidate,
+        SinglePositionRow, SinglePositionColumn, SinglePositionBlock
+        ,NakedPairBlock, NakedPairRow, NakedPairColumn
+        ,HiddenPairBlock, HiddenPairRow, HiddenPairColumn
+        ,CandidateLine
+        ]
+
+-- | Takes a sudoku and returns a solved sudoku
+solve :: Sudoku -> Sudoku -> Sudoku
+solve sud sol = applySolution sud (fst (generateSolution sud sol 0))
+
+-- | Returns how many check it took to reach a solution to a sudoku
+checksCount :: Sudoku -> Sudoku ->(Solution, Int)
+checksCount sud sol = (generateSolution sud sol 0)
+
+-- | A solver for sudoku
+applySolution :: Sudoku -> Solution -> Sudoku
+applySolution = foldl placeValueFromStep
+
+-- | Generates a Solution
+generateSolution :: Sudoku -> Sudoku -> Int -> (Solution,Int)
+generateSolution sud sol  count = case fst next of
+                            NOAVAILABLESTEPS -> ([],count)
+                            _ -> if isValid then (if count > 100000 then ([],count) else (fst next : fst res, snd res))else error "invalid solution"
+    where
+        isValid = isValid' u o 
+        next = nextStep sud (optimSteps sud) 0 --steps -- (optimSteps sud)
+        res = generateSolution (placeValueFromStep sud (fst next)) sol (count + snd next)
+        u = concat sud
+        o = concat sol
+        isValid' :: [Value] -> [Value] -> Bool
+        isValid' [] [] = True
+        isValid' (_:xs) (Empty:ys) = isValid' xs ys 
+        isValid' ((Filled x):xs) ((Filled y):ys) = and [(x == y), isValid' xs ys] 
+        isValid' (_:xs) (_:ys) = isValid' xs ys
```

Comment:
You say: typical number of needed steps:
* easy: 100-300
* difficult: 3000-10000